crates/auto-hash-map/src/map.rs

```diff
@@ -0,0 +1,276 @@
+}
+
+impl<K, V, H> Default for AutoMap<K, V, H> {
+    fn default() -> Self {
+        Self::List(Default::default())
+    }
+}
+
+impl<K: Debug, V: Debug, H> Debug for AutoMap<K, V, H> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_map().entries(self.iter()).finish()
+    }
+}
+
+impl<K, V> AutoMap<K, V, BuildHasherDefault<DefaultHasher>> {
+    pub fn new() -> Self {
+        AutoMap::List(Vec::new())
+    }
+}
+
+impl<K: Eq + Hash, V, H: BuildHasher + Default> AutoMap<K, V, H> {
+    fn convert_to_map(&mut self) -> &mut HashMap<K, V, H> {
+        if let AutoMap::List(list) = self {
+            let mut map = HashMap::with_capacity_and_hasher(MAX_LIST_SIZE * 2, Default::default());
+            for (k, v) in list.drain(..) {
+                map.insert(k, v);
+            }
+            *self = AutoMap::Map(Box::new(map));
+        }
+        if let AutoMap::Map(map) = self {
+            map
+        } else {
+            unreachable!()
+        }
+    }
+
+    fn convert_to_list(&mut self) -> &mut Vec<(K, V)> {
+        if let AutoMap::Map(map) = self {
+            let mut list = Vec::with_capacity(MAX_LIST_SIZE);
+            for (k, v) in map.drain() {
+                list.push((k, v));
+            }
+            *self = AutoMap::List(list);
+        }
+        if let AutoMap::List(list) = self {
+            list
+        } else {
+            unreachable!()
+        }
+    }
+
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+        match self {
+            AutoMap::List(list) => {
+                for (k, v) in list.iter_mut() {
+                    if *k == key {
+                        return Some(std::mem::replace(v, value));
+                    }
+                }
+                if list.len() == MAX_LIST_SIZE {
+                    let map = self.convert_to_map();
+                    map.insert(key, value);
+                } else {
+                    list.push((key, value));
+                }
+                None
+            }
+            AutoMap::Map(map) => map.insert(key, value),
+        }
+    }
+
+    pub fn remove(&mut self, key: &K) -> Option<V> {
+        match self {
+            AutoMap::List(list) => {
+                for i in 0..list.len() {
+                    if list[i].0 == *key {
+                        return Some(list.swap_remove(i).1);
+                    }
+                }
+                None
+            }
+            AutoMap::Map(map) => {
+                let result = map.remove(key);
+                if result.is_some() && map.len() < MIN_HASH_SIZE {
+                    self.convert_to_list();
+                }
+                result
+            }
+        }
+    }
+
+    pub fn entry(&mut self, key: K) -> Entry<'_, K, V> {
+        if self.len() >= MAX_LIST_SIZE {
+            self.convert_to_map();
+        } else if self.len() <= MIN_HASH_SIZE {
+            self.convert_to_list();
+        }
```

Comment:
This logic should happen in `VacantEntry::insert` and `OccupiedEntry::remove`.

Reply:
You are forcing me to write unsafe code...